### PR TITLE
feat: improvement to rpc package

### PR DESCRIPTION
**Pull request by Khaomi**
**Please describe the changes this PR makes and why it should be merged:**

**Status and versioning classification:**

### CODE REVIEW

File: packages/rpc/src/client.ts [hunk 1]
```diff
@@ -27,46 +31,66 @@ import type {
...
+	 */
 	scopes: OAuth2Scopes[];
+	/**
+	 * Token endpoint
+	 */
+	tokenEndpoint: string;
+	/**
+	 * RPC token
+	 */
 	username: string;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: Is this description correct? 🤔

Comment (Khaomi) [reply packages/rpc/src/client.ts]: i miss that, i am really sorry

File: packages/rpc/src/client.ts [hunk 2]
```diff
@@ -27,46 +31,66 @@ import type {
...
 	clientId: string;
+	/**
+	 * Client secret
+	 */
 	clientSecret: string;
-	accessToken: string;
+	/**
+	 * Should we use RPC token or npt
+	 */
 	rpcToken: string;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: This doesn't look right either

Comment (Khaomi) [reply packages/rpc/src/client.ts]: It's actually right, because rpc token is obtained from discord, not we supply it

Comment (almeidx) [reply packages/rpc/src/client.ts]: Why is the type `string` then? I don't quite understand this property / the description

Comment (Khaomi) [reply packages/rpc/src/client.ts]: because typing is weird, and the way that we assign the rpcToken required us to make it as `boolean | string`

i can fix the typing and the naming

https://discord.com/developers/docs/topics/rpc#authorize

File: packages/rpc/src/client.ts [hunk 3]
```diff
@@ -89,13 +113,18 @@ export class RPCClient extends EventEmitter {
...
+
 	private _connectPromise: Promise<unknown> | undefined;
-	private readonly _subscriptions: Map<string, ({ shortcut }: { shortcut: string }) => void>;
 
 	/**
-	 * @param {RPCClientOptions} [options] Options for the client.
+	 * @param options - Options for the client.
 	 * You must provide a transport
 	 */
 	public constructor(options: RPCClientOptions = {}) {
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: If "You must provide a transport", why is the parameter optional?

Comment (Khaomi) [reply packages/rpc/src/client.ts]: It's not optional, fixing the typing

File: packages/rpc/src/client.ts [hunk 4]
```diff
@@ -108,173 +137,183 @@ export class RPCClient extends EventEmitter {
...
+		this.fetch = async (method, path, options) => {
+			const res = await fetch(
+				`${this.endpoint}${path}${options?.query ? new URLSearchParams(options.query).toString() : ''}`,
+				{
+					method,
+					body: options?.data.toString() as string | null,
+					headers: {
+						Authorization: `Bearer ${this.accessToken!}`,
+					},
 				},
```
Comment (almeidx) [inline packages/rpc/src/client.ts:169]: Should `@discordjs/rest` be used here instead?

File: packages/rpc/src/client.ts (hunk 3 above)
Comment (almeidx) [reply packages/rpc/src/client.ts]: `You must provide a transport` can be removed from the tsdoc then

File: packages/rpc/src/client.ts [hunk 5]
```diff
@@ -27,46 +31,66 @@ import type {
...
-	return `${event}${JSON.stringify(args)}`;
-}
+import { RPCCommands, RPCEvents, type LobbyType, RelationshipType, RPCErrorCodes } from './typings/types.js';
+import { pid as getPid } from './util.js';
 
 type Class<T extends new (...args: any[]) => unknown> = T extends new (...args: any[]) => infer R ? R : never;
 
 export interface RPCClientOptions {
-	transport?: keyof typeof transports;
 	origin?: string;
```
Comment (almeidx) [inline packages/rpc/src/client.ts:52]: Missing tsdoc

File: packages/rpc/src/client.ts (hunk 4 above)
Comment (Khaomi) [reply packages/rpc/src/client.ts:169]: working on it

File: packages/rpc/src/client.ts [hunk 6]
```diff
@@ -108,173 +137,183 @@ export class RPCClient extends EventEmitter {
...
 		const data = await this.connect(options?.clientId ?? '');
 		if (!options?.scopes) {
 			this.emit(RPCEvents.Ready, data as MappedRPCDispatchData[RPCEvents.Ready]);
 			return this;
 		}
+
 		if (!options.accessToken) {
+			// eslint-disable-next-line require-atomic-updates
 			options.accessToken = await this.authorize(options);
 		}
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: Not sure if the eslint-disable comment is still required
```suggestion
		options.accessToken ??= await this.authorize(options);
```

File: packages/rpc/src/client.ts (hunk 4 above)
Comment (Khaomi) [reply packages/rpc/src/client.ts:169]: actually, porting the function to `@discordjs/rest` can be done by other contributor, i have no idea how `@discordjs/rest` work nor how error handling for it work

File: packages/rpc/src/client.ts [hunk 7]
```diff
@@ -27,46 +31,66 @@ import type {
...
-
 /**
  * The main hub for interacting with Discord RPC
- * @extends {EventEmitter}
+ *
+ * @beta
  */
 export class RPCClient extends EventEmitter {
 	public readonly options: RPCClientOptions;
+
```
Comment (almeidx) [inline packages/rpc/src/client.ts]:
```suggestion
```

File: packages/rpc/src/client.ts [hunk 8]
```diff
@@ -89,16 +113,21 @@ export class RPCClient extends EventEmitter {
...
 	/**
-	 * @param {RPCClientOptions} [options] Options for the client.
+	 * @param options - Options for the client.
 	 * You must provide a transport
 	 */
-	public constructor(options: RPCClientOptions = {}) {
+	public constructor(options: RPCClientOptions) {
 		super();
 
 		this.options = options;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]:
```suggestion
	public constructor(public readonly options: RPCClientOptions) {
		super();
```

File: packages/rpc/src/client.ts [hunk 9]
```diff
@@ -108,173 +137,183 @@ export class RPCClient extends EventEmitter {
...
 		 * Application used in this client
-		 * @type {?ClientApplication}
 		 */
 		this.application = null;
 
 		/**
 		 * User used in this application
-		 * @type {?Partial<APIUser>}
 		 */
 		this.user = null;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: If they are all just being set to null initially, you could just initialize them as null outside of the constructor (and move the tsdocs there as well)

File: packages/rpc/src/client.ts [hunk 10]
```diff
@@ -283,24 +322,26 @@ export class RPCClient extends EventEmitter {
...
-		scopes: OAuth2Scopes[];
 		clientSecret: string;
-		rpcToken: string | boolean;
-		username: string;
+		prompt?: 'consent' | 'none';
 		redirectUri?: string;
+		scopes: OAuth2Scopes[];
+		useRPCToken: boolean;
+		username: string;
 	}): Promise<string> {
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: This should be an interface

File: packages/rpc/src/client.ts [hunk 11]
```diff
@@ -469,49 +484,42 @@ export class RPCClient extends EventEmitter {
 
 	/**
 	 * Get current voice settings
-	 * @returns {Promise<GetVoiceSettingsPayloadData>}
 	 */
-	public getVoiceSettings(): Promise<GetVoiceSettingsPayloadData> {
-		return this.request(RPCCommands.GetVoiceSettings).then((s) => s as GetVoiceSettingsPayloadData);
+	public async getVoiceSettings(): Promise<GetVoiceSettingsPayloadData> {
+		return (await this.request(RPCCommands.GetVoiceSettings)) as GetVoiceSettingsPayloadData;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]:
```suggestion
		return this.request(RPCCommands.GetVoiceSettings) as Promise<GetVoiceSettingsPayloadData>;
```

File: packages/rpc/src/client.ts [hunk 12]
```diff
@@ -520,11 +528,23 @@ export class RPCClient extends EventEmitter {
 				start: activity.timestamps.start ?? Date.now(),
 				end: (activity.timestamps.end ?? null) as unknown as number,
 			};
-			if (activity.timestamps.start! > 2147483647000) {
-				throw new RangeError('timestamps.start must fit into a unix timestamp');
+
+			if (activity.timestamps.start! > 2_147_483_647_000) {
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: Can you make this a constant instead? With a descriptive name preferably

File: packages/rpc/src/client.ts [hunk 13]
```diff
@@ -620,52 +644,58 @@ export class RPCClient extends EventEmitter {
...
 		const types = Object.keys(RelationshipType);
-		return (this.request(RPCCommands.GetRelationships) as Promise<{ relationships: RelationShip[] }>).then((o) =>
-			o.relationships.map((r) => ({
-				...r,
-				type: types[r.type],
-			})),
-		);
+		const { relationships } = await (this.request(RPCCommands.GetRelationships) as Promise<{
+			relationships: RelationShip[];
+		}>);
```
Comment (almeidx) [inline packages/rpc/src/client.ts]:
```suggestion
		const { relationships } = (await this.request(RPCCommands.GetRelationships)) as <{
			relationships: RelationShip[];
		}>;
```

File: packages/rpc/src/typings/arguments.ts [hunk 14]
```diff
@@ -80,24 +80,25 @@ export interface MappedRPCCommandsArguments<SubscribeEvent extends RPCEventsWith
...
 	| SendActivityJoinInviteArguments
-	| CloseActivityRequestArguments;
+	| SetActivityArguments
+	| SetCertifiedDevicesArguments
+	| SetUserVoiceSettingsData;
 
 export interface AuthorizeArguments {
-	scopes: OAuth2Scopes[];
 	client_id: string;
+	prompt: 'consent' | 'none';
```
Comment (almeidx) [inline packages/rpc/src/typings/arguments.ts]: I think having an enum for this would be great

File: packages/rpc/src/typings/structs.ts [hunk 15]
```diff
@@ -33,23 +33,23 @@ export interface VoiceSettingDevice {
...
 
 export interface VoiceSettingsMode {
-	type: 'PUSH_TO_TALK' | 'VOICE_ACTIVITY';
 	auto_threshold: boolean;
-	threshold: number;
-	shortcut: ShortcutKeyCombo;
 	delay: number;
+	shortcut: ShortcutKeyCombo;
+	threshold: number;
+	type: 'PUSH_TO_TALK' | 'VOICE_ACTIVITY';
```
Comment (almeidx) [inline packages/rpc/src/typings/structs.ts]: Having an enum for this would also be great

File: packages/rpc/src/util.ts [hunk 16]
```diff
@@ -1,43 +1,24 @@
...
 	const { app } = require('electron');
-	// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-assignment
 	register = app.setAsDefaultProtocolClient.bind(app);
-} catch (err) {
+} catch {
 	try {
 		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-require-imports
 		register = require('register-scheme');
-	} catch (e) {} // eslint-disable-line no-empty
+	} catch {} // eslint-disable-line no-empty
```
Comment (almeidx) [inline packages/rpc/src/util.ts:13]: I believe this eslint-disable comment is not necessary
```suggestion
	} catch {}
```

**Review by almeidx — CHANGES_REQUESTED** [review]

File: packages/rpc/src/util.ts (hunk 16 above)
Comment (Khaomi) [reply packages/rpc/src/util.ts:13]: Nope, it is a problem
Screenshot 2022-10-03 at 00 50 38

File: packages/rpc/src/client.ts (hunk 5 above)
Comment (almeidx) [reply packages/rpc/src/client.ts:52]: Doesn't appear to be resolved (missing tsdoc)

File: packages/rpc/src/client.ts [hunk 17]
```diff
@@ -27,46 +31,87 @@ import type {
...
 
+const MAX_TIME_RANGE = 2_147_483_647_000;
+
+export interface AuthorizeOptions {
+	clientSecret: string;
+	prompt?: PromptType;
+	redirectUri?: string;
+	scopes: OAuth2Scopes[];
+	useRPCToken: boolean;
+	username: string;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: These should probably have tsdocs

Also, the interface looks pretty similar to `RPCLoginOptions`, perhaps it could be reused? Might have to use `Pick`/`Omit`/`Partial`

File: packages/rpc/src/client.ts [hunk 18]
```diff
@@ -27,46 +31,87 @@ import type {
...
  */
 export class RPCClient extends EventEmitter {
-	public readonly options: RPCClientOptions;
-	private accessToken: string | null;
-	public clientId: string | null;
-	public application: RPCOAuthApplication | null;
-	public user: Partial<APIUser> | null;
+	private accessToken: string | null = null;
+
+	public clientId: string | null = null;
```
Comment (almeidx) [inline packages/rpc/src/client.ts:102]: Missing tsdoc

File: .github/workflows/publish-dev.yml [hunk 19]
```diff
@@ -20,8 +20,12 @@ jobs:
             folder: 'proxy'
           - package: '@discordjs/rest'
             folder: 'rest'
+          - package: '@discordjs/rpc'
+            folder: 'rpc'
           - package: '@discordjs/util'
             folder: 'util'
+          - package: '@discordjs/rpc'
+            folder: 'rpc'
```
Comment (almeidx) [inline .github/workflows/publish-dev.yml]: Duplicate

File: packages/rpc/src/client.ts [hunk 20]
```diff
@@ -27,46 +31,87 @@ import type {
...
+	scopes: OAuth2Scopes[];
+	useRPCToken: boolean;
+	username: string;
+}
+
 export interface RPCClientOptions {
-	transport?: keyof typeof transports;
 	origin?: string;
+	/**
+	 * RPC transport. one of `ipc` or `websocket`
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: Is the `one of x or y` part really need

File: packages/rpc/src/client.ts [hunk 21]
```diff
@@ -89,192 +134,192 @@ export class RPCClient extends EventEmitter {
...
 	/**
-	 * @param {RPCClientOptions} [options] Options for the client.
+	 * @param options - Options for the client.
 	 * You must provide a transport
 	 */
-	public constructor(options: RPCClientOptions = {}) {
+	public constructor(public readonly options: RPCClientOptions) {
 		super();
 
 		this.options = options;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: This is already set by tsc due to the syntax used in the constructor parameters
```suggestion
```

File: packages/rpc/src/client.ts [hunk 22]
```diff
@@ -89,192 +134,192 @@ export class RPCClient extends EventEmitter {
...
-		 * @type {?Partial<APIUser>}
-		 */
-		this.user = null;
-
-		const Transport = options.transport ? transports[options.transport] : false;
-		if (!Transport) {
-			throw new TypeError('RPC_INVALID_TRANSPORT', options.transport as undefined);
+		const transport = transports[options.transport];
+		if (!transport) {
+			const typeError = new TypeError('INVALID_TRANSPORT', options.transport as any);
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: Would be nice to have a better naming for this variable, maybe `invalidTransport`?

File: packages/rpc/src/client.ts [hunk 23]
```diff
@@ -89,192 +134,192 @@ export class RPCClient extends EventEmitter {
...
+	 *
+	 * @param clientId - Client ID to use
 	 */
-	public connect(clientId: string) {
+	public async connect(clientId: string) {
 		if (this._connectPromise) {
 			return this._connectPromise;
 		}
+
+		const err = new Error('RPC_CONNECTION_TIMEOUT');
```
Comment (almeidx) [inline packages/rpc/src/client.ts]: Would be nice to have a better naming here as well. 

Also, are all of these `captureStackTrace`s really necessary?

File: packages/rpc/src/client.ts [hunk 24]
```diff
@@ -283,24 +328,19 @@ export class RPCClient extends EventEmitter {
...
 		redirectUri,
-	}: {
-		scopes: OAuth2Scopes[];
-		clientSecret: string;
-		rpcToken: string | boolean;
-		username: string;
-		redirectUri?: string;
-	}): Promise<string> {
-		if (clientSecret && rpcToken === true) {
+		prompt,
```
Comment (almeidx) [inline packages/rpc/src/client.ts]:
```suggestion
		prompt = PromptType.Consent,
```

File: packages/rpc/src/client.ts [hunk 25]
```diff
@@ -315,6 +355,7 @@ export class RPCClient extends EventEmitter {
 			client_id: this.clientId!,
 			rpc_token: rpcToken as string,
 			username,
+			prompt: prompt ?? 'consent',
```
Comment (almeidx) [inline packages/rpc/src/client.ts]:
```suggestion
			prompt,
```

File: packages/rpc/src/client.ts [hunk 26]
```diff
@@ -332,52 +373,53 @@ export class RPCClient extends EventEmitter {
...
+	 * @internal
+	 */
+	protected async authenticate(accessToken: string): Promise<RPCClient> {
+		const res = await (this.request(RPCCommands.Authenticate, {
+			access_token: accessToken,
+		}) as Promise<RPCAuthenticateResponsePayload>);
+
+		this.accessToken = accessToken;
+		this.application = res.data.application;
+		this.user = res.data.user;
```
Comment (almeidx) [inline packages/rpc/src/client.ts]:
```suggestion
		const { data } = (await this.request(RPCCommands.Authenticate, {
			access_token: accessToken,
		})) as RPCAuthenticateResponsePayload;

		this.accessToken = accessToken;
		this.application = data.application;
		this.user = data.user;
```

File: packages/rpc/src/client.ts [hunk 27]
```diff
@@ -469,49 +483,42 @@ export class RPCClient extends EventEmitter {
...
-		const stop = () => {
-			this._subscriptions.delete(subid);
-			return this.request(RPCCommands.CaptureShortcut, { action: 'STOP' });
-		};
-		this._subscriptions.set(subid, ({ shortcut }) => {
-			callback(shortcut, stop);
-		});
-		return this.request(RPCCommands.CaptureShortcut, { action: 'START' }).then(() => stop);
+	public async startCapturingShortcut() {
+		return this.request(RPCCommands.CaptureShortcut, { action: 'START' });
```
Comment (almeidx) [inline packages/rpc/src/client.ts:502]: Another enum for the `action`? 😄

File: packages/rpc/src/transports/ipc.ts [hunk 28]
```diff
@@ -1,11 +1,12 @@
...
+import fs from 'node:fs';
+import net, { type Socket } from 'node:net';
+import process from 'node:process';
 import type { RPCClient } from '../client';
 import type { RPCPayload, RPCResponsePayload } from '../typings/payloads';
-import { RPCCommands, RPCEvents } from '../typings/types';
-import { uuid } from '../util';
+import type { Transport } from './index';
 
 enum OPCodes {
```
Comment (almeidx) [inline packages/rpc/src/transports/ipc.ts:11]: Can you make these enum members PascalCase to match the other ones in the codebase?

File: packages/rpc/src/transports/ipc.ts [hunk 29]
```diff
@@ -15,49 +16,59 @@ enum OPCodes {
 	PONG,
 }
 
-function getIPCPath(id: number) {
+export const getIPCPath = (id: number) => {
```
Comment (almeidx) [inline packages/rpc/src/transports/ipc.ts]: Would it be a problem to make this function async in order to use `fs/promises`?

File: packages/rpc/src/typings/arguments.ts (hunk 14 above)
Comment (almeidx) [reply packages/rpc/src/typings/arguments.ts]: This should be using the PromptType enum

**Review by almeidx — CHANGES_REQUESTED** [review]

File: packages/rpc/src/client.ts (hunk 27 above)
Comment (Khaomi) [reply packages/rpc/src/client.ts:502]: I don't think you need it since it's used intenrally only

File: packages/rpc/src/transports/ipc.ts (hunk 28 above)
Comment (Khaomi) [reply packages/rpc/src/transports/ipc.ts:11]: I don't understand waht you mean

File: packages/rpc/src/client.ts (hunk 23 above)
Comment (Khaomi) [reply packages/rpc/src/client.ts]: Yes, it would give more accurate error

File: packages/rpc/CHANGELOG.md [hunk 30]
```diff
@@ -0,0 +1,11 @@
...
+
+All notable changes to this project will be documented in this file.
+
+# [unreleased]
+
+## Features
+
+- Improvment to RPC package ([62f10ce](https://github.com/discordjs/discord.js/commit/62f10cefaec2df524d755f8f3f76c957dabfaee5))
+- **rpc:** Initial commit ([17695ef](https://github.com/discordjs/discord.js/commit/17695ef698f45679e5e56393a4155c77d07ab678))
+- **rpc:** Initial commit ([ee9f53d](https://github.com/discordjs/discord.js/commit/ee9f53df3b58bbe5501c8767881f33f9707b7537))
```
Comment (imranbarbhuiya) [inline packages/rpc/CHANGELOG.md]:
```suggestion
# Changelog

All notable changes to this project will be documented in this file.
```

It'll be auto generated.

File: packages/rpc/src/transports/ipc.ts (hunk 28 above)
Comment (imranbarbhuiya) [reply packages/rpc/src/transports/ipc.ts:11]: From `HANDSHAKE` to `Handshake` like this

File: packages/rpc/src/transports/websocket.ts [hunk 31]
```diff
@@ -1,21 +1,25 @@
...
+import { browser } from '..';
 import type { RPCClient } from '../client';
-import { browser } from '../index';
+import type { Transport } from './index';
 
 type WebSocketConstructor = new (url: string, options?: WebSocketType.ClientOptions) => WebSocketType;
 
-// @ts-expect-error
-// eslint-disable-next-line
+// @ts-expect-error - TS doesn't know that window can exist if browser is true
```
Comment (imranbarbhuiya) [inline packages/rpc/src/transports/websocket.ts]: Why not just add `dom` lib?

File: packages/rpc/tsup.config.ts [hunk 32]
```diff
@@ -1,3 +1,3 @@
-import { createTsupConfig } from '../../tsup.config';
```
Comment (imranbarbhuiya) [inline packages/rpc/tsup.config.ts:1]: Other packages now uses `.js` so u can change the file name to `.js`

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/rpc/src/transports/ipc.ts (hunk 28 above)
Comment (Khaomi) [reply packages/rpc/src/transports/ipc.ts:11]: Ohh, i was looking at the wrong file, my bad.

File: packages/rpc/.prettierignore [hunk 33]
```diff
@@ -5,5 +5,4 @@ dist/
 docs/**/*
 !docs/index.yml
 !docs/README.md
-coverage/
-tsup.config.*.mjs
+coverage/
```
Comment (almeidx) [inline packages/rpc/.prettierignore]: Missing new line at the end of the file

File: packages/rpc/src/client.ts [hunk 34]
```diff
@@ -27,46 +31,86 @@ import type {
...
+	 */
 	clientId: string;
+	/**
+	 * Client secret
+	 */
 	clientSecret: string;
-	accessToken: string;
-	rpcToken: string;
-	tokenEndpoint: string;
+	prompt: PromptType;
```
Comment (almeidx) [inline packages/rpc/src/client.ts:72]: Missing tsdoc

**Review by almeidx — CHANGES_REQUESTED** [review]

File: packages/rpc/src/client.ts (hunk 34 above)
Comment (Khaomi) [reply packages/rpc/src/client.ts:72]: I am not really good with writing docs, is it possible for other contributor to work on documentation?

**Review by almeidx — APPROVED** [review]

File: packages/rpc/README.md [hunk 35]
```diff
@@ -1,93 +1,73 @@
...
 ```
 
 ## Links
 
 - [Website](https://discord.js.org/) ([source](https://github.com/discordjs/discord.js/tree/main/packages/website))
-- [Documentation](https://discord.js.org/#/docs/rest)
+- [Documentation](https://discord.js.org/#/docs/rpc)
 - [Guide](https://discordjs.guide/) ([source](https://github.com/discordjs/guide))
-  See also the [Update Guide](https://discordjs.guide/additional-info/changes-in-v13.html), including updated and removed items in the library.
+  See also the [Update Guide](https://discordjs.guide/additional-info/changes-in-v14.html), including updated and removed items in the library.
```
Comment (megatank58) [inline packages/rpc/README.md:60]: The update guide does not has rpc information so this is somewhat redunant

File: packages/rpc/package.json [hunk 36]
```diff
@@ -52,24 +53,26 @@
...
+		"eslint": "^8.24.0",
+		"eslint-config-neon": "^0.1.33",
 		"prettier": "^2.7.1",
-		"tsup": "^6.1.3",
-		"typescript": "^4.7.4",
-		"vitest": "^0.17.0"
+		"tsup": "^6.2.3",
+		"typescript": "^4.8.4",
+		"vitest": "^0.23.4"
 	},
```
Comment (megatank58) [inline packages/rpc/package.json]: are all these dependencies such as undici really needed?

File: packages/rpc/src/client.ts [hunk 37]
```diff
@@ -27,46 +31,86 @@ import type {
...
+	 */
+	public clientId: string | null = null;
+
+	/**
+	 * Application used in this client
+	 */
+	public application: RPCOAuthApplication | null = null;
+
+	/**
+	 * User used in this application
```
Comment (megatank58) [inline packages/rpc/src/client.ts:110]: maybe rephrase this to `User using the rpc` or something else, the user isn't used in the application but uses the application

File: packages/rpc/src/error.ts [hunk 38]
```diff
@@ -0,0 +1,18 @@
+import { RPCErrorCodes } from './typings/types.js';
+
+export class RPCError extends Error {
+	public code: RPCErrorCodes;
+
+	public override message: string = '';
+
+	public override get name() {
+		return `RpcError${RPCErrorCodes[this.code]}`;
```
Comment (megatank58) [inline packages/rpc/src/error.ts]:
```suggestion
		return `RpcError ${RPCErrorCodes[this.code]}`;
```
wouldn't it be better have `RpcError 42069` than `RpcError42069`

**Comment by megatank58** [comment]
some documentation comments would also probably need to be rewritten

**Comment by Khaomi** [comment]
I will let other contributor work on docs because i am really bad at writing docs.

File: packages/rpc/package.json (hunk 36 above)
Comment (Khaomi) [reply packages/rpc/package.json]: we are applying changes from upstream / other packages

File: packages/rpc/README.md (hunk 35 above)
Comment (Khaomi) [reply packages/rpc/README.md:60]: we are applying changes from upstream / other packages
